### PR TITLE
HIVE-26838: Added support for reload event in HMS

### PR DESCRIPTION
**Pull request by saihemanth-cloudera**
### What changes were proposed in this pull request?
Added support for a new event in HMS.

### Why are the changes needed?
This will improve cache performance in external systems using HMS for metadata operations.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Remote cluster.

### CODE REVIEW

**Comment by saihemanth-cloudera** [comment]
@dengzhhu653 @nrg4878 Can you please review this PR?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/ReloadEvent.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+     * @param status status of insert, true = success, false = failure
+     * @param refreshEvent status of insert,
+     * @param handler handler that is firing the event
+     */
+    public ReloadEvent(String catName, String db, String table, List<String> partVals, boolean status,
+                       boolean refreshEvent, Map<String, String> tblParams, IHMSHandler handler) throws MetaException,
+            NoSuchObjectException {
+        super(status, handler);
+
+        GetTableRequest req = new GetTableRequest(db, table);
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/ReloadEvent.java:60]: Why do we need to fetch the table/partitions given that this is a refresh or invalidate command? For refreshEvent = false, is `db` or `table` nullable?

I wondering how the client(impala) suppose to use this? do they receive the notification and extract the `table` or `partition` from the message, or do they request `HMS` for the meta?

File: standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift [hunk 2]
```diff
@@ -1473,9 +1473,11 @@ struct InsertEventRequestData {
 }
 
 union FireEventRequestData {
-    1: InsertEventRequestData insertData
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:1476]: nit: we need to update [hive_metastore.proto]( https://github.com/apache/hive/blob/master/standalone-metastore/metastore-common/src/main/protobuf/org/apache/hadoop/hive/metastore/hive_metastore.proto) as well...
https://cwiki.apache.org/confluence/display/Hive/HiveDeveloperFAQ#HiveDeveloperFAQ-Howtogenerateprotobufcode?

File: standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/GetFileMetadataRequest.java [hunk 3]
```diff
@@ -396,13 +396,13 @@ public void write(org.apache.thrift.protocol.TProtocol prot, GetFileMetadataRequ
     public void read(org.apache.thrift.protocol.TProtocol prot, GetFileMetadataRequest struct) throws org.apache.thrift.TException {
       org.apache.thrift.protocol.TTupleProtocol iprot = (org.apache.thrift.protocol.TTupleProtocol) prot;
       {
-        org.apache.thrift.protocol.TList _list1017 = iprot.readListBegin(org.apache.thrift.protocol.TType.I64);
-        struct.fileIds = new java.util.ArrayList<java.lang.Long>(_list1017.size);
-        long _elem1018;
-        for (int _i1019 = 0; _i1019 < _list1017.size; ++_i1019)
+        org.apache.thrift.protocol.TList _list1027 = iprot.readListBegin(org.apache.thrift.protocol.TType.I64);
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/GetFileMetadataRequest.java:399]: nit: do you know why changes happen here? given that we only change `FireEventRequestData` in the `hive_metastore.thrift`.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/ReloadEvent.java (hunk 1 above)
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/ReloadEvent.java:60]: The client will use this table/partition object from the event for further processing. The Client will check if the table in the event is up-to-date using the latest notification id of the table, if so it'll directly use this table object from the event, or else it'll fetch from HMS.

File: standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/GetFileMetadataRequest.java (hunk 3 above)
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/GetFileMetadataRequest.java:399]: These changes occur with the change in the FireEventRequest object. I did a lot of tests around this on why so many thrift files are changing when I changed the FireEventRequest object but I couldn't figure out a reason.

File: standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift (hunk 2 above)
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:1481]: ack

File: standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/GetFileMetadataRequest.java (hunk 3 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/GetFileMetadataRequest.java:399]: The result shows no diff when I complied the thrift with this PR.

**Review by dengzhhu653 — APPROVED** [review]

**Pull request merged**